### ISSUE TITLE
Bootswatch Flatly under Plugins

### ISSUE THREAD

**Issue by gtbu** [Typesetter/Typesetter#334]
After installation of TP 511b1 of 02-12-2019 the Bootswatch Flatly appears under Plugins
![flatplug](https://user-images.githubusercontent.com/30436307/69961921-2ce22e00-150d-11ea-83a3-fbd216648b83.jpg)

**Comment by juek** [comment]
Already discussed [here](https://github.com/Typesetter/Typesetter/issues/323#issuecomment-552002722)

**Comment by gtbu** [comment]
Discusses  as appeasring in the CE Version - but to this is 5.11b1 ! - no excuse.....

**Comment by mahotilo** [comment]
I do not have such problem.

**Comment by juek** [comment]
Behavior changed with [this commit](https://github.com/Typesetter/Typesetter/commit/1b94b5f4bd5d1333534df1d5890e074deaf79012).
Still subject to change, though.

**Comment by juek** [comment]
FYI: the change might need a clean install in order to get rid of already registered hook entries.

**Comment by juek** [comment]
@mahotilo 

can you imagine any useful filtering or action mechanisms to be implemented in the FrontEndFramework ini key, which would call anything dynamically at runtime? 
I cannot at them moment.

**Comment by mahotilo** [comment]
```
[FrontEndFramework]
name = Bootstrap
version = 3
```
OK, now it is hint for used version of Bootstrap. But it is hardcoded in \gp\tool::LoadComponents() in template.php.

Maybe we can put in this section a list of connected gadgets, if they are hardcoded in template? Notifying users that they need to install additional plugins for used theme can be useful.

**Comment by juek** [comment]
Just to clarify the terminology: Typesetter 'gadgets' cannot be part of a theme, while plugins hooks can. 

If I understand you correctly, what you mean would be a different key, sth. like

[Dependencies]
plugins = 'Simple Blog, Catalog Easy'

Correct?

**Comment by mahotilo** [comment]
Yes,
speaking (writing) of  gadgets, I meant something like this in template.php
```
		<div class="col-md d-none d-md-block">		
			<?php gpOutput::Get('Gadget','BBC_Online_Visitor') ?>
			<?php gpOutput::GetAdminLink();?></div>
		</div>

```

**Comment by juek** [comment]
Unfortunately it's getting a bit off-topic, so I'll make it short.

I've already done that and it's anything but trivial. My current solution does not rely on an Addon.ini entry but solves all sorts of dependencies using a custom 'post-installer' in the theme. It is not yet ready for publishing but I can share a screenshot with you:

![grafik](https://user-images.githubusercontent.com/704283/69973058-1cd54900-1523-11ea-885a-48e1ef9d8648.png)

**Comment by juek** [comment]
... of course ther is has also a
![grafik](https://user-images.githubusercontent.com/704283/69973131-45f5d980-1523-11ea-966d-6cc4214b2700.png) button ;)

**Comment by mahotilo** [comment]
Wow 8)

**Comment by juek** [comment]
another sneak peek:

![grafik](https://user-images.githubusercontent.com/704283/69973815-683c2700-1524-11ea-9aa9-c6784df28e49.png)

Of course, such things have a price tag: We get the same trouble we have to tame WordPress premium themes and behave the way **we** want them. Everything with an admin interface tends to be more complex on the programmatic side.

**Comment by mahotilo** [comment]
Although I am not a fan of themes that can be customized via interface, positively it is cool and a great work.

**Comment by juek** [comment]
> Although I am not a fan of themes that can be customized via interface…

So am I. But people without programming skills need them. Current Bootwatch themes are great boilerplates for developers, but 'ordinary' people are having a hard time even adding a brand logo.

**Comment by mahotilo** [comment]
Agree, but how demanding are 'ordinary' people to the template. They will always be unhappy with the available settings...
Off-topic, off-topic, off-topic:) 
Great job! It will be appreciated.

**Comment by juek** [comment]
True.

@gtbu 

if the original issue is solved from your POV, can we close?

**Comment by gtbu** [comment]
Though i do not want to ride on it ...(fresh installation of the master)
![flatplug2](https://user-images.githubusercontent.com/30436307/69994291-dba65e80-154d-11ea-8515-875aef9f27b7.jpg)

**Comment by juek** [comment]
Thx - seems as I have to check it again.

> I hope its not windows and Laragon and php 7.2

Certainly not.

**Comment by juek** [comment]
Should now be fixed with [this commit](https://github.com/Typesetter/Typesetter/commit/95a38e869689a0fa7c3db51764b76d590a29a2c7). 

To unregister as plugin, all layouts deriving from themes formerly registered as plugin need to be 'deleted' via 
Appearance -> Manage -> Bootswatch Flatly/LayoutName -> Layout Options->Delete 
and then reactivated via 
Appearance -> Manage ->Available

**Comment by gtbu** [comment]
A new installation under laragon delivers after fresh installation the error :  does not exist. Would you like to create it? SCSS Compile Failed: parse error: failed at `$icon-font-path: 
"../../include/thirdparty/Bootstrap3/fonts/";` (stdin) on line �

**Comment by juek** [comment]
30 seconds is way too long for a local development server to compile Bootswatch Flatly (Less). 
My Raspberry Pi 3 only takes 5-6 seconds for that task. My 'real' local server (8-core Xeon) does it in less than 1s. Doesn't speak for Laragon IMO.

**Comment by gtbu** [comment]
May i allow me inspite of poor my i5-3570 and 4 virusscanners  to offer You this file for test
typesetter5.bplaced.net/data/_uploaded/file/less.php.zip
Just delete the less.php-directory and fill it with my bad but fast code - hopefully - Yamantaka will like You then !

**Comment by juek** [comment]
Compile Bootswatch Flatyl/Sticky Footer Less

Test results on Raspberry Pi 3:

Current less.php (TS 5.1.1-b1):

1.19 MB Memory
13.15 MB Max Memory
4.714 Seconds
4714 Milliseconds

Your version:

1.02 MB Memory
12.51 MB Max Memory
3.616 Seconds
3616 Milliseconds

**Comment by gtbu** [comment]
This measurement - I do not know the procedure - takes place on a 'dedicated server'. On Xampp and Laragon I have a factor of at least 5  --- in a fast web of about 2. People with small ram have bigger problems. On the github page of Ojeyorge is also referred to the further development in the wiki less-version. But heck - my corresponding CE version is 2.53 and correspondingly fast and extensive, and  I do not want to force anyone to happiness.

**Comment by mahotilo** [comment]
> May i allow me inspite of my poor i5-3570 and 4 virusscanners to offer You this file for test

I use only one simple MSE antivirus on my i5, but I exclude dir with my local web-server form checking due to its noticeable impact on site speed.
Could it be, that your CE version is detected by system/scanners as local files and checked only once, but TS downloaded from github is detected like as a dangerous object for permanent monitoring? Could this be a source of speed loss? Have you checked TS-b2 without 4 virusscanners?

**Comment by gtbu** [comment]
I have Avira free, Iobit malware fighter 7.3 with bitdefender (giveawayoftheday.com) and Secure aplus  with 12 engines(share on sale) - thats the reason for speed loss inspite of 16 GB Ram : The CE-Version is quite the same - just version 1.83 and with a different folder and less2.php (and php7.4 ready). My scanners check everything ... just not spyare (adwcleaner 8 is for that!).  The version of the above link is 1.7 as the momentaray version of tp511b1 but  abit faster. The new version 1.8 ist not tested with php5.4 (Hm - YOU can do that - it is quite easy !)  Seen from a long distance You will have to use the wiki-version anyway (if not less-i - which need bigger adaption)

**Comment by juek** [comment]
> a 'dedicated server'

Although the RasPi acts as a dedicated server (it hasn't much else to do normally), I'd say a 3rd-gen i5 is roughly 30 to 40 times faster than the RasPi's ARMv8, and a recent Xeon some 200 times. 
Your issue with local windows dev hosts is most likely due to antivirus stepping in. Especially when large amounts of temp files are created, which is the case upon less compilation. Not a real-world issue IMO.

**Comment by juek** [comment]
> Seen from a long distance You will have to use the wiki-version anyway

I personally do not use less anymore, since Scss is more powerful and the compiler performs better.

**Comment by juek** [comment]
> This measurement - I do not know the procedure

In this case I simply used the Admin Menu -> Performance section. 
With a fresh setup without installed plugins…

* assign BS Flatly to a single page, use a BS Scss theme as standard. 
* clear the cache (Tools -> Resource Cache).
* return to a standard page and let Typesetter recompile the Scss and Admin part, then switch to your BS Flatly page. 
* when the page shows up (after the Less compilation finished) the Performance section will show the results, mostly caused by the Less compiler.

You can subtract some 20-40 ms and ~0.7 MB memory usage used for the normal CMS operation.

**Comment by gtbu** [comment]
Sorrey - but i disagree : I prefer, after complete installation without Flatly 1.02 themes and then cleaned cache , the choice of a flatly thme, then click at it and measure the time until it completely appears in the 'chose as template'.  This delay is the nerving point in some webinstalltions. 
The wiki-versions is remarkable faster : In the normal shared hosting the memory is 32 (east europe, cheap and free providers) to 256 MB (Strato,1&1) and a delay comes from harddisk-stepping and queue with other users. So we have a worst case scenaro and an average scenario and to consider poor installations. But as i said -  I do not want to force anyone to happiness - i am like Yamantaka....

**Comment by juek** [comment]
In order to clarify my point of view of which LESS compiler we use at the moment - for everyone reading along.

Josh stopped the maintenance of [his less.php compiler](https://github.com/oyejorge/less.php) some time ago. [Wikimedia forked](https://github.com/wikimedia/less.php) and still actively develops it further, which is fine. 
In the meantime Bootstrap also [offiicially moved](https://blog.getbootstrap.com/2015/08/19/bootstrap-4-alpha/) from LESS to Scss, same as Typesetter. We have ScssPHP for quite some years, which is IMO superior for several reasons. However Josh's less.php still is part of Typesetter and required to process legacy Bootswatch Flatly (and others) less-based themes. 

Wikimedia's current less.php version requires PHP 7.2.9. Typesetter still supports PHP 5. This is why we will not simply upgrade, which would lock out quite some older Typesetter installations out of our current bug fixes.

With the next major release, whenever it comes, things will look different.

**Comment by gtbu** [comment]
No - its not quite like that . Josh's version was still under vendor at mediawiki 1.29 and now removed. 
Verion 1.80 [should be ok ](https://packagist.org/packages/wikimedia/less.php). Mediawiki got the versions from[ github-asenar](https://github.com/Asenar/less.php/releases). Version 1.80 shpould be 5.3 to 7.3. For users of php7.3 version 1.7.010 has some bugs.

**Comment by juek** [comment]
Well, if …

* it works with our current minimum system requirements (PHP 5.4 but I could also live with 5.6 for Typesetter 5.2) and 
* it's actually faster, which appears to be so

… we should use it. 

But we should then strip it down to what is actually needed. Your package is ~2.3 MB in size while our current version is only 256 KB.

@oyejorge Mind to look into this?

**Comment by oyejorge** [comment]
Will do!

**Comment by juek** [comment]
Great!

I can confirm that it works with PHP 5.4.4

For testing please update to 
#343 and this [commit](https://github.com/Typesetter/Typesetter/commit/20e6b6f2f2d038ad0761db1177d112658af1ab97)
